### PR TITLE
Added JSDocs for a lot of classes

### PR DESCRIPTION
**Pull request by Harbs**
I added JSDocs and type information for a lot of classes. It's not complete, but it's a very good start.

You can view some generated JSDocs [on my site here.](http://in-tools.com/stuff/opentype.js/)

### CODE REVIEW

**Comment by oriSomething** [comment]
wow 👍

**Comment by Harbs** [comment]
On second thought, I'm not sure the JSDocs that I did are totally the right way to go. I needed to do some non-standard tags to make it work and exporting Google externs does not work very well. You will also notice that trying to view the source from the JSDocs does not work, and I believe it's casued by the same problem.

(I'm also not sure I exported all type info correctly.)

If there's any chance of it being merged, I'll work on a bit of a structure refactor to work with JSDocs a bit better.

Thoughts?

File: src/font.js [hunk 1]
```diff
@@ -9,9 +9,42 @@ var glyphset = require('./glyphset');
...
+ * @property {string=} license
+ * @property {string=} licenseURL
+ * @property {string=} version
+ * @property {string=} description
+ * @property {string=} copyright
+ * @property {string=} trademark
+ * @property {Number} unitsPerEm
+ * @property {Number} ascender
+ * @property {Number} descender
+ * @property {Boolean=} createdTimestamp
```
Comment (Jolg42) [inline src/font.js]: createdTimestamp is a number 😉

File: src/types.js [hunk 2]
```diff
@@ -394,13 +549,18 @@ encode.MACSTRING = function(str, encoding) {
                 return undefined;
             }
         }
-
-        result.push(c);
+        result[i] = c;
```
Comment (Jolg42) [inline src/types.js]: Do you know @fpirsch if ` result.push(c);` is faster than direct assignment `result[i] = c;`?
It looks like it's faster… and easier to read too :) http://stackoverflow.com/a/15650115/1345244

Comment (Harbs) [inline src/types.js]: Here's a test that I did in Firefox and Chrome with results:
function timeJoin(){
	var start = new Date();
	var i=-1;
	var len  = 10000000;
	var arr = [];
	while(++i<len)
	{
		arr[i] = "a";
	}
	var str = arr.join("");
	var end = new Date();
	console.log(end.getTime()-start.getTime());
};
function timePush(){
	var start = new Date();
	var i=-1;
	var len  = 10000000;
	var arr = [];
	while(++i<len)
	{
		arr.push("a");
	}
	var str = arr.join("");
	var end = new Date();
	console.log(end.getTime()-start.getTime());
};

Chrome:
timeJoin()
822
timePush()
979

Firefox:
timeJoin()
184
timePush()
303

Unless there's something wrong with my test, push seems to be slower -- at least in Firefox. (Safari seems to be slightly slower with push as well.) Older browsers had a much bigger gap. I'm not sure about IE.

**Comment by Jolg42** [comment]
Hi @Harbs 
It looks good 👍  I think we can merge that soon, more documentation is always nice!

I just think that the changes from '.push()' to direct assignment is less readable & looks not as fast from a performance view, @fpirsch will know that better ;)

I don't know if it's possible to take the file "opentype.js/src/opentype.js" as the entry of JSDoc, because it will be better to have the real source when clicking on the generated doc (not the dist/opentype.js), & it will be definitely faster too.

File: src/types.js (hunk 2 above)
Comment (Jolg42) [inline src/types.js]: @Harbs After searching for the best one in term of performance it looks like it's more complicated than that…

> Usually there is no reason not to use push - the method is there for exactly that operation and makes some code easier to read. While a few people think one version is faster than the other, both are equally optimized in browsers. See Why is array.push sometimes faster than array[n] = value? and Using the push method or .length when adding to array? - results vary so wide that it's actually irrelevant. Use what is better to understand.

from http://stackoverflow.com/a/21034712/1345244

> Actually push is just as fast as property assignment. Good measurements are rare, but those that are done properly show varying results across different browser engine versions - changing rapidly and unexpected. See Appending to array, Why is array.push sometimes faster than array[n] = value? and Is there a reason JavaScript developers don't use Array.push()? - the conclusion is that you should use what is most readable / appropriate for your use case, not what you think could be faster.
from http://stackoverflow.com/a/18477047/1345244

Also, I know that AirBnB did a good job with this public style guide https://github.com/airbnb/javascript/blob/master/es5/README.md#arrays and use `.push()`.
In the end maybe the performance is not really easy to compare here and readability is more important?

**Comment by fpirsch** [comment]
The fastest way is to pre-allocate the array with `new Array(length)`
What is slow is changing the size of arrays.
I did not have a close look but it seems to me that in this method the array length is known from the start, so it could be preallocated.

**Comment by Jolg42** [comment]
@fpirsch 100% for pre-allocating the arrays 👍  but for direct vs push it looks more complicated no?

**Comment by fpirsch** [comment]
@Jolg42 
If the array is preallocated, you can't use `push` anyway ;-)

@Harbs 
> On second thought, I'm not sure the JSDocs that I did are totally the right way to go.

JSDoc is basically a port of JavaDoc, but JS is not Java. JS is not about classes and inheritance, so in most practical cases JSDoc is awkward. Anyway JSDoc is better than no doc at all :-)

This is why I only use basic JSDoc syntax, it's a simple and familiar way to document things, but not intended to compile a real JSDoc (which would, btw, be mostly useless).

**Comment by Harbs** [comment]
Cool.

Feel free to revert my changes vis a vis push(), but push() is almost always slower in most browsers.

About JSDoc: I would like to restructure the modules a bit to make JSDoc and exporting Google externs cleaner. externs are pretty important for using the library in strongly typed environments. If that's okay with you, I'll work on that.

It will not be major restructuring, but there's certain structures in the code that does not work so well with JSDoc and extern tools.

**Comment by Harbs** [comment]
BTW, re. array performance, it's likely that using TypedArrays for a lot of the calculations would be more performant and less memory intensive.

**Comment by Jolg42** [comment]
So in the end we will keep your array changes and I will check one day for non pre-allocated arrays ;)
And yes TypedArrays can be great at performance and already used in the lib for some parts 👍

**Comment by fpirsch** [comment]
My experience with typed arrays has always been very disappointing with regards to speed in V8 (Chrome, Node). But I did not check recently if things have changed.

**Comment by Harbs** [comment]
Here's an interesting SO discussion on the topic: http://stackoverflow.com/questions/24853686/javascript-typedarray-performance Both answers are enlightening.

BTW, I regenerated the JSDocs using the original src files, and the results are much better.
http://in-tools.com/stuff/opentype.js/

I'm also still looking into how to structure the code so externs can be generated correctly.

**Comment by fdb** [comment]
@Harbs this looks very good, thanks! 

I would like to see one improvement: keep all the direct assignment / vs push stuff for a separate pull request (I prefer if possible to pre-allocate the array and use direct assignment as @fpirsch mentioned).

**Pull request merged**